Fix ARMv7 build warnings

justinmichaud opened the pull request:
#### 356b22df6a2afa45dc9f47eaafa5c21d57ee51db
<pre>
Fix ARMv7 build warnings
<a href="https://bugs.webkit.org/show_bug.cgi?id=288678">https://bugs.webkit.org/show_bug.cgi?id=288678</a>

Reviewed by Darin Adler.

There are many cast alignment warnings that we skip on 32-bit, mostly related
to GObject* downcasts. Since these warnings aren&apos;t very helpful, we just disable
them on 32-bit.

We also fix a few uint64_t &lt;&gt; size_t confusions with format strings.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
* Source/WebCore/platform/graphics/opentype/OpenTypeMathData.cpp:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreTile.cpp:
(WebCore::CoordinatedBackingStoreTile::processPendingUpdates):
* Source/WebCore/platform/libwpe/PlatformPasteboardLibWPE.cpp:
(WebCore::PlatformPasteboard::getTypes const):
(WebCore::PlatformPasteboard::readString const):
* Source/WebKit/UIProcess/API/C/wpe/WKPagePrivateWPE.cpp:
(WKPageHandleKeyboardEvent):
* Source/cmake/WebKitCommon.cmake:
* Tools/MiniBrowser/wpe/main.cpp:
* Tools/Scripts/webkitdirs.pm:
(generateBuildSystemFromCMakeProject):

Canonical link: <a href="https://commits.webkit.org/292361@main">https://commits.webkit.org/292361@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/891b7a43e9cb7eb647907332b50c7e466acc8bee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95739 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15347 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5262 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100790 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46244 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97783 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15635 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23780 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73014 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30261 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98742 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11704 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86469 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53348 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11429 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4197 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45583 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/88407 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81599 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4307 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102824 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/94359 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22797 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16623 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82056 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23049 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82496 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81420 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20417 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25986 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3445 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16134 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22765 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27914 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/117838 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22424 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25900 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24166 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/cmake/WebKitCommon.cmake:
```diff
@@ -100,7 +100,11 @@ if (NOT HAS_RUN_WEBKIT_COMMON)
     endif ()
     if (LOWERCASE_CMAKE_SYSTEM_PROCESSOR MATCHES "(^aarch64|^arm64|^cortex-?[am][2-7][2-8])")
-        set(WTF_CPU_ARM64 1)
+         if (FORCE_32BIT)
```
This will go away once rebased

justinmichaud commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -2844,6 +2844,16 @@ sub generateBuildSystemFromCMakeProject
         $ENV{"LDFLAGS"} = "-m32" . ($ENV{"LDFLAGS"} || "");
     }
+    if (architecture() eq "arm64" && shouldBuild32Bit()) {
+        # CMAKE_LIBRARY_ARCHITECTURE is needed to get the right .pc
```
^ same

philn commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreTile.cpp:
```diff
@@ -48,13 +48,13 @@ void CoordinatedBackingStoreTile::processPendingUpdates(TextureMapper& textureMa
         return;
 
-    WTFBeginSignpost(this, CoordinatedSwapBuffers, "%lu updates", updatesCount);
+    WTFBeginSignpost(this, CoordinatedSwapBuffers, "%zu updates", updatesCount);
```
Would it make more sense to use the PRui64 (not sure this is the right one, always getting confused with the letters) formatters?

aperezdc replied:
Here `"%zu"` is the correct format because the variable has type `size_t`, which might be 32-bit in some architectures. Using `"%"PRIu64` should be done only when the variable is definitely known to be 64-bit wide like `uint64_t`.

Constellation replied:
PRInt64 makes code super hard to read. And even it is not stable for various other type usage like size_t. I would like to just have "%llu" and cast uint64_t in the parameter.

justinmichaud replied:
Unfortunately it still has to be llu on 32-bit, and lu on 64-bit. So we have to choose between a cast + PRIu64, or zu

aperezdc replied:
Note that `%zu` will automatically match the width-size of a `size_t` for the current platform, no need for `PRIxxx` macros, nor to choose between `%lu` and `%llu`. Rule of thumb: use always `%zu` for `size_t` and `%zd` for `ssize_t`.

philn commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -40,7 +41,9 @@
 #endif
 
 #if !USE_GSTREAMER_FULL && (ENABLE_WEB_AUDIO || ENABLE_VIDEO)
+WTF_IGNORE_WARNINGS_IN_THIRD_PARTY_CODE_BEGIN
```
what is the issue here?

justinmichaud replied:
Internal GObject* casting issues

mcatanzaro commented on Source/WebCore/CMakeLists.txt:
```diff
@@ -2608,6 +2608,12 @@ if (PORT STREQUAL GTK OR PORT STREQUAL WPE)
 
+if (ARM_THUMB2_DETECTED)
+    WEBKIT_ADD_TARGET_CXX_FLAGS(WebCore
+        -Wno-cast-align
```
Documentation of this warning:

```
       -Wcast-align
           Warn whenever a pointer is cast such that the required alignment of the target is increased.  For example, warn if a "char *" is cast to an "int
           *" on machines where integers can only be accessed at two- or four-byte boundaries.
```

Are you sure it's safe to turn this off? Looks like it indicates serious bugs? We could alternatively use `-Wcast-align=strict` to surface the warnings on other architectures so they're not limited to just 32-bit ARM.

justinmichaud replied:
The problem is that GObject is 4-byte aligned, but the second you put anything interesting inside a struct on armv7, it becomes 8-byte aligned.

I would love to have a more robust solution to this, but we can't exactly make GObject* declare its alignment to be bigger. Unfortunately I also don't know of a way to force a greater alignment for imported types.

I suppose we could define a helper to do these downcasts for us, and have it squash this warning (or ASSERT the correct alignment), would that be better? It would probably churn a lot of code.

cast-align=strict seems like a good option if we make that change.

mcatanzaro replied:
I don't know what we should do about this.

darinadler replied:
I think the real question is not about the _warning_, but about why the alignment mistake is not a real problem. If it is a real problem, then we can start using a separate memory block for the things that need more alignment instead of putting the data members directly into the class derived from `GObject`. If it’s not a problem it would be good to understand better why, and then I _do_ think the helper functions would be a good strategy to avoid turning off the warning for all the nearby code that might benefit from it just to quiet the.

eugeneia replied:
What is the failure mode? So if we ever allocated a GObject at 4 byte aligned address, and then cast it to a subclass that needs to be 8 byte aligned, and we'd access a member, we'd likely receive a SIGBUS and be killed?

> but we can't exactly make GObject* declare its alignment to be bigger

Why not? That seems to me like the elegant solution.

mcatanzaro replied:
You can't change GObject ABI ever. That's one level short of changing glibc ABI; think great social upheaval, riots in the streets, possibly the end of times. (OK, I exaggerate a little. _A little._) Got to just make the best of the way things are.

> What is the failure mode? So if we ever allocated a GObject at 4 byte aligned address, and then cast it to a subclass that needs to be 8 byte aligned, and we'd access a member, we'd likely receive a SIGBUS and be killed?

I'm not sure; I've never actually understood alignment requirements. But GObjects get cast between types very frequently, so that sounds like the sort of thing that would happen constantly, and we're not seeing bug reports about SIGBUS on ARM. Then again, probably there are very few users testing WebKitGTK on 32-bit ARM.

darinadler replied:
We don’t need to debate this here, clearly the WebKit project is not in change of glib, but for the record, this is absolutely not proposing “changing GObject ABI”, what we are proposing is correctly annotating it so the compiler correctly understands alignment that is the actual GObject ABI. The fact that they are always allocated aligned is part of the ABI! We’ve tackled so many of these types problems in the past on other platforms (both Apple ones and others); this is very familiar territory for the project. To make these warnings work properly in WebKit we should figure out the most expedient approach; it would be great to get the warnings for all the non-GObject things, so it would be good to find a way to deal with it without turning off the warning entirely.

mcatanzaro replied:
Pretty sure changing the alignment (e.g. by using `_Alignas`) would be an ABI break? It's not hard to make changes in GLib, so long as those changes don't break the ABI. But if the change would require rebuilding software that depends on GLib, the change will be rejected.

darinadler replied:
Like I said, this is not where we’d make those decisions. I’m pretty sure such a change would not require rebuilding.

Constellation commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreTile.cpp:
```diff
@@ -48,13 +48,13 @@ void CoordinatedBackingStoreTile::processPendingUpdates(TextureMapper& textureMa
             continue;
 
-        WTFBeginSignpost(this, CoordinatedSwapBuffer, "%u/%lu, rect %ix%i+%i+%i", updateIndex + 1, updatesCount, update.tileRect.x(), update.tileRect.y(), update.tileRect.width(), update.tileRect.height());
+        WTFBeginSignpost(this, CoordinatedSwapBuffer, "%u/%zu, rect %ix%i+%i+%i", updateIndex + 1, updatesCount, update.tileRect.x(), update.tileRect.y(), update.tileRect.width(), update.tileRect.height());
```
Ditto

darinadler commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreTile.cpp:
```diff
@@ -44,15 +44,15 @@ void CoordinatedBackingStoreTile::addUpdate(Update&& update)
 {
     auto updates = WTFMove(m_updates);
-    auto updatesCount = updates.size();
+    uint64_t updatesCount = updates.size();
```
```suggestion
    size_t updatesCount = updates.size();
```
Or we can leave it as `auto`. No need to switch to 64-bit.

darinadler commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreTile.cpp:
```diff
@@ -44,15 +44,15 @@ void CoordinatedBackingStoreTile::addUpdate(Update&& update)
         return;
 
-    WTFBeginSignpost(this, CoordinatedSwapBuffers, "%lu updates", updatesCount);
+    WTFBeginSignpost(this, CoordinatedSwapBuffers, "%" PRIu64 " updates", updatesCount);
```
```suggestion
    WTFBeginSignpost(this, CoordinatedSwapBuffers, "%zu updates", updatesCount);
```

darinadler commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreTile.cpp:
```diff
@@ -44,15 +44,15 @@ void CoordinatedBackingStoreTile::addUpdate(Update&& update)
         auto& update = updates[updateIndex];
 
-        WTFBeginSignpost(this, CoordinatedSwapBuffer, "%u/%lu, rect %ix%i+%i+%i", updateIndex + 1, updatesCount, update.tileRect.x(), update.tileRect.y(), update.tileRect.width(), update.tileRect.height());
+        WTFBeginSignpost(this, CoordinatedSwapBuffer, "%u/%" PRIu64 ", rect %ix%i+%i+%i", updateIndex + 1, updatesCount, update.tileRect.x(), update.tileRect.y(), update.tileRect.width(), update.tileRect.height());
```
```suggestion
        WTFBeginSignpost(this, CoordinatedSwapBuffer, "%u/%zu, rect %ix%i+%i+%i", updateIndex + 1, updatesCount, update.tileRect.x(), update.tileRect.y(), update.tileRect.width(), update.tileRect.height());
```

darinadler commented on Source/WebCore/dom/SpaceSplitString.cpp:
```diff
@@ -195,7 +195,7 @@ inline Ref<SpaceSplitStringData> SpaceSplitStringData::create(const AtomString&
 
     RELEASE_ASSERT(tokenCount < (std::numeric_limits<unsigned>::max() - sizeof(SpaceSplitStringData)) / sizeof(AtomString));
-    unsigned sizeToAllocate = sizeof(SpaceSplitStringData) + tokenCount * sizeof(AtomString);
+    ptrdiff_t sizeToAllocate = sizeof(SpaceSplitStringData) + tokenCount * sizeof(AtomString);
```
Why not `size_t`? I don't see anything here that's signed.

Why keep the assertion above that this fits in an `unsigned` if that's no longer used?

justinmichaud replied:
This was just because the assertion at the bottom compares it with a ptrdiff_t, but I agree that it makes more sense to change the assertion

justinmichaud committed:
Fix ARMv7 build warnings
https://bugs.webkit.org/show_bug.cgi?id=288678

Reviewed by Darin Adler.

There are many cast alignment warnings that we skip on 32-bit, mostly related
to GObject* downcasts. Since these warnings aren't very helpful, we just disable
them on 32-bit.

We also fix a few uint64_t <> size_t confusions with format strings.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
* Source/WebCore/platform/graphics/opentype/OpenTypeMathData.cpp:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreTile.cpp:
(WebCore::CoordinatedBackingStoreTile::processPendingUpdates):
* Source/WebCore/platform/libwpe/PlatformPasteboardLibWPE.cpp:
(WebCore::PlatformPasteboard::getTypes const):
(WebCore::PlatformPasteboard::readString const):
* Source/WebKit/UIProcess/API/C/wpe/WKPagePrivateWPE.cpp:
(WKPageHandleKeyboardEvent):
* Source/cmake/WebKitCommon.cmake:
* Tools/MiniBrowser/wpe/main.cpp:
* Tools/Scripts/webkitdirs.pm:
(generateBuildSystemFromCMakeProject):

Canonical link: https://commits.webkit.org/292361@main
